packages/react-dom/src/server/ReactPartialRenderer.js

```diff
@@ -682,6 +682,9 @@ class ReactDOMServerRenderer {
     context: Object,
     parentNamespace: string,
   ): string {
+    if (typeof element.type !== 'string') {
+      invariant(false, 'ReactElement type is invalid -- expected a string');
```

Comment:
Not necessarily. *This function* expected a string. But that's just because we let the code go far enough to go into this function. From user's point of view, they might expect it to be a custom component but they messed up the import and got `undefined` instead.

We should check earlier and show a full message, with the same wording as on the client. We should also add a test case to `ReactDOMIntegrationElement-test` similar to existing `itThrowsWhenRendering` calls in it. This way we verify the behavior is identical on the client and on the server.

Reply:
@gaearon should [this](https://github.com/facebook/react/blob/master/packages/react/src/ReactElementValidator.js#L280) then be an `invariant` and have a shorter version of this error thrown when `NODE_ENV=production`? Otherwise, production would have the `toLowerCase` error displayed

Reply:
No, what I'm saying is we should check if it's a string just before entering `renderDOM`. If it's not a string then we should throw.

I guess technically it means the check *could* be inside `renderDOM` but I still find the flow to be more confusing since it depends on `renderDOM` call being the last branch. Moving the check out would make more sense to me.

Reply:
Ah I see in `render` which in turn calls `renderDOM` has lots of checks in it. I will update PR soon!